AsyncSnapshotingTest.shouldSucceedToTakeSnapshotOnNextIntervalWhenLastWritePosRetrievingFailed

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? [At least once](https://ci.zeebe.camunda.cloud/blue/organizations/jenkins/camunda-cloud%2Fzeebe/detail/release-1.4.0-alpha1/1/tests)
- Does it block your work? No
- Do we suspect that it is a real failure? I don't think so

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>

```
org.awaitility.core.ConditionTimeoutException: 
Assertion condition defined as a io.camunda.zeebe.broker.system.partitions.AsyncSnapshotingTest 
Expecting Optional to contain a value but it was empty within 10 seconds.
	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:165)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:895)
	at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:679)
	at io.camunda.zeebe.broker.system.partitions.AsyncSnapshotingTest.shouldSucceedToTakeSnapshotOnNextIntervalWhenLastWritePosRetrievingFailed(AsyncSnapshotingTest.java:175)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.lambda$execute$1(JUnitPlatformProvider.java:199)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:193)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:120)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: java.lang.AssertionError: 
Expecting Optional to contain a value but it was empty.
	at io.camunda.zeebe.broker.system.partitions.AsyncSnapshotingTest.lambda$shouldSucceedToTakeSnapshotOnNextIntervalWhenLastWritePosRetrievingFailed$5(AsyncSnapshotingTest.java:175)
	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:222)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:209)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

<details><summary>Logs</summary>
<pre>

```
Standard Output
12:55:36.922 [] [main] DEBUG io.camunda.zeebe.logstreams.snapshot - Opened database from '/tmp/junit5834930714341401105/runtime'.
12:55:37.132 [Broker-1-SnapshotStore-1] [-zb-fs-workers-1] DEBUG io.camunda.zeebe.logstreams.snapshot - Taking temporary snapshot into /tmp/junit5834930714341401105/pending/125-1-25-9223372036854775807.
12:55:37.234 [Broker-1-SnapshotStore-1] [-zb-fs-workers-0] DEBUG io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot - Aborting transient snapshot FileBasedTransientSnapshot{directory=/tmp/junit5834930714341401105/pending/125-1-25-9223372036854775807, checksum=2956754977, metadata=FileBasedSnapshotMetadata{index=125, term=1, processedPosition=25, exporterPosition=9223372036854775807}}
12:55:37.157 [Broker-0-SnapshotDirector-1] [-zb-fs-workers-1] ERROR io.camunda.zeebe.logstreams.snapshot - Unexpected error in resolving last written position.
java.lang.RuntimeException: getLastWrittenPositionAsync fails
	at io.camunda.zeebe.broker.system.partitions.AsyncSnapshotingTest.shouldSucceedToTakeSnapshotOnNextIntervalWhenLastWritePosRetrievingFailed(AsyncSnapshotingTest.java:160) ~[test-classes/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.RunRules.evaluate(RunRules.java:20) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42) ~[junit-vintage-engine-5.8.2.jar:5.8.2]
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80) ~[junit-vintage-engine-5.8.2.jar:5.8.2]
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72) ~[junit-vintage-engine-5.8.2.jar:5.8.2]
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53) ~[junit-platform-launcher-1.8.2.jar:1.8.2]
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.lambda$execute$1(JUnitPlatformProvider.java:199) ~[surefire-junit-platform-3.0.0-M5.jar:3.0.0-M5]
	at java.util.Iterator.forEachRemaining(Iterator.java:133) ~[?:?]
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:193) ~[surefire-junit-platform-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154) ~[surefire-junit-platform-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:120) ~[surefire-junit-platform-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428) ~[surefire-booter-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162) ~[surefire-booter-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562) ~[surefire-booter-3.0.0-M5.jar:3.0.0-M5]
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548) ~[surefire-booter-3.0.0-M5.jar:3.0.0-M5]
12:55:47.355 [] [main] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-fs-workers'
12:55:47.355 [] [main] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-actors'
12:55:47.356 [] [-zb-fs-workers-1] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-fs-workers': closed successfully
12:55:47.360 [] [-zb-actors-4] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-actors': closed successfully
12:55:47.472 [] [main] DEBUG io.camunda.zeebe.logstreams.snapshot - Closed database from '/tmp/junit5834930714341401105/runtime'.
```

</pre>
</details>

--- a/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java
@@ -27,7 +27,6 @@ import io.camunda.zeebe.util.sched.future.CompletableActorFuture;
 import java.time.Duration;
 import java.util.HashSet;
 import java.util.Map;
-import java.util.Optional;
 import java.util.Set;
 import java.util.function.BooleanSupplier;
 import org.slf4j.Logger;
@@ -58,7 +57,7 @@ public final class AsyncSnapshotDirector extends Actor
   private Long lastWrittenPosition;
   private TransientSnapshot pendingSnapshot;
   private long lowerBoundSnapshotPosition;
-  private CompletableActorFuture<Optional<PersistedSnapshot>> snapshotFuture;
+  private CompletableActorFuture<PersistedSnapshot> snapshotFuture;
   private boolean persistingSnapshot;
 
   @SuppressWarnings("java:S3077") // allow volatile here, health is immutable
@@ -194,11 +193,11 @@ public final class AsyncSnapshotDirector extends Actor
   /**
    * Directly take a snapshot, independently of the scheduled snapshots.
    *
-   * @return A future that is completed successfully when the snapshot was taken or if the snapshot
-   *     was skipped.
+   * @return A future that is completed successfully when the snapshot was taken. If the snapshot
+   *     was skipped, the future is also completed successfully but with a null.
    */
-  public CompletableActorFuture<Optional<PersistedSnapshot>> forceSnapshot() {
-    final var newSnapshotFuture = new CompletableActorFuture<Optional<PersistedSnapshot>>();
+  public CompletableActorFuture<PersistedSnapshot> forceSnapshot() {
+    final var newSnapshotFuture = new CompletableActorFuture<PersistedSnapshot>();
     actor.call(() -> prepareTakingSnapshot(newSnapshotFuture));
     return newSnapshotFuture;
   }
@@ -219,10 +218,10 @@ public final class AsyncSnapshotDirector extends Actor
   }
 
   private void prepareTakingSnapshot(
-      final CompletableActorFuture<Optional<PersistedSnapshot>> newSnapshotFuture) {
+      final CompletableActorFuture<PersistedSnapshot> newSnapshotFuture) {
     if (snapshotFuture != null) {
       LOG.debug("Already taking snapshot, skipping this request for a new snapshot");
-      newSnapshotFuture.complete(Optional.empty());
+      newSnapshotFuture.complete(null);
       return;
     }
 
@@ -235,7 +234,7 @@ public final class AsyncSnapshotDirector extends Actor
             if (lastProcessedPosition == StreamProcessor.UNSET_POSITION) {
               LOG.debug(
                   "We will skip taking this snapshot, because we haven't processed something yet.");
-              snapshotFuture.complete(Optional.empty());
+              snapshotFuture.complete(null);
               snapshotFuture = null;
               return;
             }
@@ -343,7 +342,7 @@ public final class AsyncSnapshotDirector extends Actor
                 LOG.error(ERROR_MSG_MOVE_SNAPSHOT, persistError);
               }
             } else {
-              snapshotFuture.complete(Optional.of(snapshot));
+              snapshotFuture.complete(snapshot);
             }
             lastWrittenPosition = null;
             snapshotFuture = null;

--- a/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java
@@ -13,6 +13,7 @@ import io.camunda.zeebe.broker.system.partitions.StateController;
 import io.camunda.zeebe.engine.processing.streamprocessor.StreamProcessor;
 import io.camunda.zeebe.engine.processing.streamprocessor.StreamProcessorMode;
 import io.camunda.zeebe.logstreams.impl.Loggers;
+import io.camunda.zeebe.snapshots.PersistedSnapshot;
 import io.camunda.zeebe.snapshots.SnapshotException;
 import io.camunda.zeebe.snapshots.SnapshotException.SnapshotNotFoundException;
 import io.camunda.zeebe.snapshots.TransientSnapshot;
@@ -26,6 +27,7 @@ import io.camunda.zeebe.util.sched.future.CompletableActorFuture;
 import java.time.Duration;
 import java.util.HashSet;
 import java.util.Map;
+import java.util.Optional;
 import java.util.Set;
 import java.util.function.BooleanSupplier;
 import org.slf4j.Logger;
@@ -56,7 +58,7 @@ public final class AsyncSnapshotDirector extends Actor
   private Long lastWrittenPosition;
   private TransientSnapshot pendingSnapshot;
   private long lowerBoundSnapshotPosition;
-  private boolean takingSnapshot;
+  private CompletableActorFuture<Optional<PersistedSnapshot>> snapshotFuture;
   private boolean persistingSnapshot;
 
   @SuppressWarnings("java:S3077") // allow volatile here, health is immutable
@@ -124,7 +126,7 @@ public final class AsyncSnapshotDirector extends Actor
         snapshotRate,
         failure);
 
-    resetStateOnFailure();
+    resetStateOnFailure(failure);
     healthReport = HealthReport.unhealthy(this).withIssue(failure);
 
     for (final var listener : listeners) {
@@ -133,7 +135,7 @@ public final class AsyncSnapshotDirector extends Actor
   }
 
   /**
-   * Create an AsyncSnapshotDirector that can take snapshot when the Streamprocessor is in
+   * Create an AsyncSnapshotDirector that can take snapshot when the StreamProcessor is in
    * continuous replay mode.
    *
    * @param nodeId id of this broker
@@ -185,12 +187,20 @@ public final class AsyncSnapshotDirector extends Actor
   }
 
   private void scheduleSnapshotOnRate() {
-    actor.runAtFixedRate(snapshotRate, this::prepareTakingSnapshot);
-    prepareTakingSnapshot();
+    actor.runAtFixedRate(snapshotRate, () -> prepareTakingSnapshot(new CompletableActorFuture<>()));
+    prepareTakingSnapshot(new CompletableActorFuture<>());
   }
 
-  public void forceSnapshot() {
-    actor.call(this::prepareTakingSnapshot);
+  /**
+   * Directly take a snapshot, independently of the scheduled snapshots.
+   *
+   * @return A future that is completed successfully when the snapshot was taken or if the snapshot
+   *     was skipped.
+   */
+  public CompletableActorFuture<Optional<PersistedSnapshot>> forceSnapshot() {
+    final var newSnapshotFuture = new CompletableActorFuture<Optional<PersistedSnapshot>>();
+    actor.call(() -> prepareTakingSnapshot(newSnapshotFuture));
+    return newSnapshotFuture;
   }
 
   @Override
@@ -208,12 +218,15 @@ public final class AsyncSnapshotDirector extends Actor
     actor.run(() -> listeners.remove(failureListener));
   }
 
-  private void prepareTakingSnapshot() {
-    if (takingSnapshot) {
+  private void prepareTakingSnapshot(
+      final CompletableActorFuture<Optional<PersistedSnapshot>> newSnapshotFuture) {
+    if (snapshotFuture != null) {
+      LOG.debug("Already taking snapshot, skipping this request for a new snapshot");
+      newSnapshotFuture.complete(Optional.empty());
       return;
     }
 
-    takingSnapshot = true;
+    snapshotFuture = newSnapshotFuture;
     final var futureLastProcessedPosition = streamProcessor.getLastProcessedPositionAsync();
     actor.runOnCompletion(
         futureLastProcessedPosition,
@@ -222,7 +235,8 @@ public final class AsyncSnapshotDirector extends Actor
             if (lastProcessedPosition == StreamProcessor.UNSET_POSITION) {
               LOG.debug(
                   "We will skip taking this snapshot, because we haven't processed something yet.");
-              takingSnapshot = false;
+              snapshotFuture.complete(Optional.empty());
+              snapshotFuture = null;
               return;
             }
 
@@ -231,7 +245,8 @@ public final class AsyncSnapshotDirector extends Actor
 
           } else {
             LOG.error(ERROR_MSG_ON_RESOLVE_PROCESSED_POS, error);
-            takingSnapshot = false;
+            snapshotFuture.completeExceptionally(error);
+            snapshotFuture = null;
           }
         });
   }
@@ -247,7 +262,7 @@ public final class AsyncSnapshotDirector extends Actor
             } else {
               LOG.error("Failed to take a snapshot for {}", processorName, snapshotTakenError);
             }
-            resetStateOnFailure();
+            resetStateOnFailure(snapshotTakenError);
             return;
           }
 
@@ -271,7 +286,7 @@ public final class AsyncSnapshotDirector extends Actor
       persistingSnapshot = false;
       persistSnapshotIfLastWrittenPositionCommitted();
     } else {
-      resetStateOnFailure();
+      resetStateOnFailure(error);
       LOG.error(ERROR_MSG_ON_RESOLVE_WRITTEN_POS, error);
     }
   }
@@ -318,6 +333,7 @@ public final class AsyncSnapshotDirector extends Actor
       snapshotPersistFuture.onComplete(
           (snapshot, persistError) -> {
             if (persistError != null) {
+              snapshotFuture.completeExceptionally(persistError);
               if (persistError instanceof SnapshotNotFoundException) {
                 LOG.warn(
                     "Failed to persist transient snapshot {}. Nothing to worry if a newer snapshot exists.",
@@ -326,18 +342,24 @@ public final class AsyncSnapshotDirector extends Actor
               } else {
                 LOG.error(ERROR_MSG_MOVE_SNAPSHOT, persistError);
               }
+            } else {
+              snapshotFuture.complete(Optional.of(snapshot));
             }
             lastWrittenPosition = null;
-            takingSnapshot = false;
+            snapshotFuture = null;
             pendingSnapshot = null;
             persistingSnapshot = false;
           });
     }
   }
 
-  private void resetStateOnFailure() {
+  private void resetStateOnFailure(final Throwable failure) {
+    if (snapshotFuture != null && !snapshotFuture.isDone()) {
+      snapshotFuture.completeExceptionally(failure);
+    }
+
     lastWrittenPosition = null;
-    takingSnapshot = false;
+    snapshotFuture = null;
     if (pendingSnapshot != null) {
       pendingSnapshot.abort();
       pendingSnapshot = null;

--- a/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/AsyncSnapshottingTest.java
+++ b/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/AsyncSnapshottingTest.java
@@ -133,8 +133,8 @@ public final class AsyncSnapshottingTest {
     createAsyncSnapshotDirectorOfProcessingMode();
     final var firstSnapshot = asyncSnapshotDirector.forceSnapshot();
     setCommitPosition(99L);
-    assertThat(firstSnapshot.join()).isPresent();
-    final var firstSnapshotIndex = firstSnapshot.join().orElseThrow().getIndex();
+    assertThat(firstSnapshot.join()).isNotNull();
+    final var firstSnapshotIndex = firstSnapshot.join().getIndex();
 
     // when
     final var secondSnapshot = asyncSnapshotDirector.forceSnapshot();
@@ -143,10 +143,9 @@ public final class AsyncSnapshottingTest {
     // then
     assertThat(secondSnapshot.join())
         .describedAs("Second snapshot is taken")
-        .isPresent()
-        .get()
-        .extracting(PersistedSnapshot::getIndex, as(InstanceOfAssertFactories.LONG))
+        .isNotNull()
         .describedAs("Second snapshot has a higher index")
+        .extracting(PersistedSnapshot::getIndex, as(InstanceOfAssertFactories.LONG))
         .isGreaterThan(firstSnapshotIndex);
     assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
   }
@@ -202,7 +201,7 @@ public final class AsyncSnapshottingTest {
     setCommitPosition(commitPosition);
 
     // then
-    assertThat(secondSnapshot.join()).isPresent();
+    assertThat(secondSnapshot.join()).isNotNull();
     assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
   }
 
@@ -213,7 +212,7 @@ public final class AsyncSnapshottingTest {
     final var snapshot = asyncSnapshotDirector.forceSnapshot();
 
     // then
-    assertThat(snapshot.join()).isPresent();
+    assertThat(snapshot.join()).isNotNull();
     assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
   }
 }

--- a/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/AsyncSnapshottingTest.java
+++ b/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/AsyncSnapshottingTest.java
@@ -7,7 +7,9 @@
  */
 package io.camunda.zeebe.broker.system.partitions;
 
+import static org.assertj.core.api.Assertions.as;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.timeout;
@@ -30,14 +32,14 @@ import java.io.IOException;
 import java.time.Duration;
 import java.util.Optional;
 import org.agrona.concurrent.UnsafeBuffer;
-import org.awaitility.Awaitility;
+import org.assertj.core.api.InstanceOfAssertFactories;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.RuleChain;
 import org.junit.rules.TemporaryFolder;
 
-public final class AsyncSnapshotingTest {
+public final class AsyncSnapshottingTest {
 
   private final TemporaryFolder tempFolderRule = new TemporaryFolder();
   private final AutoCloseableRule autoCloseableRule = new AutoCloseableRule();
@@ -93,7 +95,8 @@ public final class AsyncSnapshotingTest {
         .thenReturn(CompletableActorFuture.completed(32L));
 
     when(mockStreamProcessor.getLastWrittenPositionAsync())
-        .thenReturn(CompletableActorFuture.completed(99L), CompletableActorFuture.completed(100L));
+        .thenReturn(CompletableActorFuture.completed(99L))
+        .thenReturn(CompletableActorFuture.completed(100L));
   }
 
   private void createAsyncSnapshotDirectorOfProcessingMode() {
@@ -114,37 +117,38 @@ public final class AsyncSnapshotingTest {
   public void shouldValidSnapshotWhenCommitPositionGreaterEquals() {
     // given
     createAsyncSnapshotDirectorOfProcessingMode();
-    asyncSnapshotDirector.forceSnapshot();
+    final var snapshot = asyncSnapshotDirector.forceSnapshot();
 
     // when
     setCommitPosition(100L);
 
     // then
-    Awaitility.await()
-        .untilAsserted(() -> assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent());
+    assertThat(snapshot.join()).isNotNull();
+    assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
   }
 
   @Test
   public void shouldTakeSnapshotsOneByOne() {
     // given
     createAsyncSnapshotDirectorOfProcessingMode();
-    asyncSnapshotDirector.forceSnapshot();
+    final var firstSnapshot = asyncSnapshotDirector.forceSnapshot();
     setCommitPosition(99L);
-    Awaitility.await()
-        .untilAsserted(() -> assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent());
-    final PersistedSnapshot oldSnapshot = persistedSnapshotStore.getLatestSnapshot().orElseThrow();
+    assertThat(firstSnapshot.join()).isPresent();
+    final var firstSnapshotIndex = firstSnapshot.join().orElseThrow().getIndex();
 
     // when
-    asyncSnapshotDirector.forceSnapshot();
+    final var secondSnapshot = asyncSnapshotDirector.forceSnapshot();
     setCommitPosition(100L);
 
     // then
-    Awaitility.await()
-        .untilAsserted(
-            () ->
-                assertThat(persistedSnapshotStore.getCurrentSnapshotIndex())
-                    .describedAs("New snapshot is taken")
-                    .isGreaterThan(oldSnapshot.getIndex()));
+    assertThat(secondSnapshot.join())
+        .describedAs("Second snapshot is taken")
+        .isPresent()
+        .get()
+        .extracting(PersistedSnapshot::getIndex, as(InstanceOfAssertFactories.LONG))
+        .describedAs("Second snapshot has a higher index")
+        .isGreaterThan(firstSnapshotIndex);
+    assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
   }
 
   @Test
@@ -157,22 +161,21 @@ public final class AsyncSnapshotingTest {
 
     when(mockStreamProcessor.getLastProcessedPositionAsync())
         .thenReturn(CompletableActorFuture.completed(lastProcessedPosition));
+    final var initialFailure = new RuntimeException("getLastWrittenPositionAsync fails");
     when(mockStreamProcessor.getLastWrittenPositionAsync())
-        .thenReturn(
-            CompletableActorFuture.completedExceptionally(
-                new RuntimeException("getLastWrittenPositionAsync fails")));
+        .thenReturn(CompletableActorFuture.completedExceptionally(initialFailure));
     setCommitPosition(commitPosition);
-    asyncSnapshotDirector.forceSnapshot();
-    verify(mockStreamProcessor, timeout(5000).times(1)).getLastWrittenPositionAsync();
+    assertThatThrownBy(() -> asyncSnapshotDirector.forceSnapshot().join()).hasCause(initialFailure);
+    verify(mockStreamProcessor, timeout(10000).times(1)).getLastWrittenPositionAsync();
 
     // when
     when(mockStreamProcessor.getLastWrittenPositionAsync())
         .thenReturn(CompletableActorFuture.completed(lastWrittenPosition));
-    asyncSnapshotDirector.forceSnapshot();
 
     // then
-    Awaitility.await()
-        .untilAsserted(() -> assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent());
+    assertThat(asyncSnapshotDirector.forceSnapshot().join()).isNotNull();
+    assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
+    verify(mockStreamProcessor, timeout(10000).times(2)).getLastWrittenPositionAsync();
   }
 
   @Test
@@ -183,34 +186,34 @@ public final class AsyncSnapshotingTest {
     final long lastWrittenPosition = 26L;
     final long commitPosition = 100L;
 
+    final var initialFailure = new RuntimeException("getLastProcessedPositionAsync fails");
     when(mockStreamProcessor.getLastProcessedPositionAsync())
-        .thenReturn(
-            CompletableActorFuture.completedExceptionally(
-                new RuntimeException("getLastProcessedPositionAsync fails")));
+        .thenReturn(CompletableActorFuture.completedExceptionally(initialFailure));
     when(mockStreamProcessor.getLastWrittenPositionAsync())
         .thenReturn(CompletableActorFuture.completed(lastWrittenPosition));
-    asyncSnapshotDirector.forceSnapshot();
+
+    assertThatThrownBy(() -> asyncSnapshotDirector.forceSnapshot().join()).hasCause(initialFailure);
     verify(mockStreamProcessor, timeout(5000).times(1)).getLastProcessedPositionAsync();
 
     // when
     when(mockStreamProcessor.getLastProcessedPositionAsync())
         .thenReturn(CompletableActorFuture.completed(lastProcessedPosition));
-    asyncSnapshotDirector.forceSnapshot();
+    final var secondSnapshot = asyncSnapshotDirector.forceSnapshot();
     setCommitPosition(commitPosition);
 
     // then
-    Awaitility.await()
-        .untilAsserted(() -> assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent());
+    assertThat(secondSnapshot.join()).isPresent();
+    assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
   }
 
   @Test
   public void shouldPersistSnapshotWithoutWaitingForCommitWhenInReplayMode() {
     // when
     createAsyncSnapshotDirectorOfReplayMode();
-    asyncSnapshotDirector.forceSnapshot();
+    final var snapshot = asyncSnapshotDirector.forceSnapshot();
 
     // then
-    Awaitility.await()
-        .untilAsserted(() -> assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent());
+    assertThat(snapshot.join()).isPresent();
+    assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
   }
 }

--- a/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/AsyncSnapshottingTest.java
+++ b/broker/src/test/java/io/camunda/zeebe/broker/system/partitions/AsyncSnapshottingTest.java
@@ -124,7 +124,7 @@ public final class AsyncSnapshottingTest {
 
     // then
     assertThat(snapshot.join()).isNotNull();
-    assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
+    assertThat(persistedSnapshotStore.getLatestSnapshot()).hasValue(snapshot.join());
   }
 
   @Test
@@ -147,7 +147,7 @@ public final class AsyncSnapshottingTest {
         .describedAs("Second snapshot has a higher index")
         .extracting(PersistedSnapshot::getIndex, as(InstanceOfAssertFactories.LONG))
         .isGreaterThan(firstSnapshotIndex);
-    assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
+    assertThat(persistedSnapshotStore.getLatestSnapshot()).hasValue(secondSnapshot.join());
   }
 
   @Test
@@ -202,7 +202,7 @@ public final class AsyncSnapshottingTest {
 
     // then
     assertThat(secondSnapshot.join()).isNotNull();
-    assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
+    assertThat(persistedSnapshotStore.getLatestSnapshot()).hasValue(secondSnapshot.join());
   }
 
   @Test
@@ -213,6 +213,6 @@ public final class AsyncSnapshottingTest {
 
     // then
     assertThat(snapshot.join()).isNotNull();
-    assertThat(persistedSnapshotStore.getLatestSnapshot()).isPresent();
+    assertThat(persistedSnapshotStore.getLatestSnapshot()).hasValue(snapshot.join());
   }
 }

